Add Volume Health Information

tiagolobocastro opened the pull request:
refactor(pstor/etcd-watcher): remove watch status update
    
    The keep alive is already useful for triggering watch restarts in case the
    connection is in a zombie state.
    Also, to address what happens when no progress status is received within
    the timeout, we can also issue an explicit request for progress and update
    the internal rev this way. This is also a much safer way of updating the
    rev as there might some race between getting the status and watches?
    
---

    feat(pstor/etcd): add http2 keep-alives
    
    This helps recover from half open connections automatically.
    
---

    test(etcd): allow for persistent data
    
    Allows testing etcd restarts, whilst persisting its data.
    
---

    refactor: add new volume health info keys
    
    Use new types for volume health info and info prefix, rather than
    modify the existing NexusInfoKey.
    We should also split the NexusInfoKey into non-volume info keys (which
    are only used for testing anyway) and volume info keys.
    
---

    security(GHSA-4p46-pwfr-66x6): update ring dependency
    
    advisory: https://github.com/advisories/GHSA-4p46-pwfr-66x6
    
---

    feat(health): add knob to disable volume health usage
    
---

    test(volume/watcher): add test for the various conditions
    
    Checks the volume health as replicas are faulted, rebuilt and
    targets shutdown cleanly or otherwise.

---

    fix(rest-plugin): show missing target if missing
    
    If the target is meant to be there, but it's missing, state that.
    
---

    feat(rest-plugin): add clean shutdown to volume output
    
---

    feat(rest-plugin): add replica topology health info
    
---

    feat(rest): add health information
    
    Enhance the volume with pstor health information which is retried from
    etcd using watches.
    
---

    feat(core/watcher): add pstor watcher for volume health info
    
    Add a volume health watcher which caches updates from etcd for every volume.
    When etcd entry is deleted, the equivalent entry in the cache is also deleted.
    
    When volume is republished sucessfully, the older target health information are
    also removed from the cache.
    
---

    fix: remove volume health entries on delete
    
    When a volume is deleted, destroy all related health entries.
    It has been observed that upon republished, the previous info
    is not deleted, so even if we fix that a lot of entries may
    already be leaked, so let's ensure we delete all entries for
    a given volume on volume deletion.
    
---

    refactor(pstor/watcher): add missing value on watch callback

tiagolobocastro committed:
refactor(pstor/watcher): add missing value on watch callback

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix: remove volume health entries on delete

When a volume is deleted, destroy all related health entries.
It has been observed that upon republished, the previous info
is not deleted, so even if we fix that a lot of entries may
already be leaked, so let's ensure we delete all entries for
a given volume on volume deletion.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(core/watcher): add pstor watcher for volume health info

Add a volume health watcher which caches updates from etcd for every volume.
When etcd entry is deleted, the equivalent entry in the cache is also deleted.

When volume is republished sucessfully, the older target health information are
also removed from the cache.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(rest): add health information

Enhance the volume with pstor health information which is retried from
etcd using watches.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(rest-plugin): add replica topology health info

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(rest-plugin): add clean shutdown to volume output

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(rest-plugin): show missing target if missing

If the target is meant to be there, but it's missing, state that.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
test(volume/watcher): add test for the various conditions

Checks the volume health as replicas are faulted, rebuilt and
targets shutdown cleanly or otherwise.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on control-plane/agents/src/bin/core/controller/pstor_cache.rs:
```diff
@@ -0,0 +1,51 @@
+};
+
+#[derive(Clone)]
+pub(crate) struct PStorCache {
```
Add a doc comment?

tiagolobocastro replied:
Done

tiagolobocastro committed:
feat(health): add knob to disable volume health usage

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
security(GHSA-4p46-pwfr-66x6): update ring dependency

advisory: https://github.com/advisories/GHSA-4p46-pwfr-66x6

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

dsharma-dc commented on control-plane/agents/src/bin/core/volume/operations.rs:
```diff
@@ -109,15 +109,7 @@ impl ResourceLifecycle for OperationGuardArc<VolumeSpec> {
-            )
-            .await;
-        }
+        self.delete_all_info(registry).await;
```
Comment needs to be updated to reflect new behaviour.
Also, suggest renaming to `delete_all_nexusinfo`.

tiagolobocastro replied:
Done

dsharma-dc commented on control-plane/stor-port/src/types/v0/store/nexus_persistence.rs:
```diff
@@ -55,6 +55,7 @@ pub struct NexusInfoKey {
     volume_id: Option<VolumeId>,
     nexus_id: NexusId,
     mayastor_compat_v1: bool,
+    rm_all_nexuses: bool,
```
This seems bit out of place on (one) NexusInfoKey while it intends to operate on volume level(all nexusinfo keys).

tiagolobocastro replied:
Yeah I wasn't sure where to place it, since the logic which builds the exact key is only contained in this type. Any suggestions?

dsharma-dc replied:
How about not setting any nexus_id while creating NexusInfoKey from `delete_all_info`? And then during key creation assume that `volume_id + no nexus_id` means delete all nexus infos for volume. AFAIK, currently the code always expects a nexus_id while volume_id is optional.

tiagolobocastro replied:
Making nexus optional here would require a lot of changes.
Instead I added new types, reused logic which builds key for the nexus info.
Please take a look.

tiagolobocastro committed:
refactor: add new volume health info keys

Use new types for volume health info and info prefix, rather than
modify the existing NexusInfoKey.
We should also split the NexusInfoKey into non-volume info keys (which
are only used for testing anyway) and volume info keys.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

dsharma-dc commented on control-plane/agents/src/bin/core/volume/registry.rs:
```diff
@@ -106,8 +112,37 @@ impl Registry {
+                } else {
+                    online_healthy_replicas.min(1)
+                },
+                live_healthy_replicas: healthy
```
what's the difference b/w `online_healthy_replicas` and `live_healthy_replicas` ?

tiagolobocastro replied:
`live_healthy_replicas` are `online_healthy_replicas` which are in use by the volume target (nexus)

dsharma-dc commented on control-plane/agents/src/bin/core/controller/resources/volume/mod.rs:
```diff
@@ -55,3 +63,93 @@ macro_rules! volume_span {
+            if arg.value.is_empty() {
+                match NexusInfoKey::parse_id(arg.updated_key) {
+                    Ok(id) => {
+                        //tracing::warn!(key=%arg.updated_key, "Removing key");
```
remove this line?

tiagolobocastro replied:
doh, indeed

dsharma-dc commented on control-plane/agents/src/bin/core/controller/registry.rs:
```diff
@@ -165,6 +168,7 @@ impl Registry {
             inner: Arc::new(RegistryInner {
                 nodes: Default::default(),
                 specs: ResourceSpecsLocked::new(),
+                health: VolumeHealthWatcher::new(&store),
```
IIUC, the first time health info is cached blank when registry initializes, i.e. in init_health. And afterwards, this watcher updates those `entries` upon any event on nexusinfo key?
What happens to updated health info upon next registry refresh(cache_period)?

tiagolobocastro replied:
On `new` it's blank but we register the watch, so any updates will start being inserted.
Then on `init_health` we read from etcd and we add to the cache only if the watch hasn't yet inserted them.

> What happens to updated health info upon next registry refresh(cache_period)?

The cache is not used by this health, health is driven by the watch only (ie no polling)

tiagolobocastro committed:
test(etcd): allow for persistent data

Allows testing etcd restarts, whilst persisting its data.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(pstor/etcd): add http2 keep-alives

This helps recover from half open connections automatically.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
refactor(pstor/etcd-watcher): remove watch status update

The keep alive is already useful for triggering watch restarts in case the
connection is in a zombie state.
Also, to address what happens when no progress status is received within
the timeout, we can also issue an explicit request for progress and update
the internal rev this way. This is also a much safer way of updating the
rev as there might some race between getting the status and watches?

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge
